[libcxx][docs] Make test name pattern documentation more obvious

hawkinsw opened the pull request:
As a new contributor, I found it hard to find the documentation for the meaning of the names of different tests and how those names translate to Lit.

This makes that documentation more explicit.

hawkinsw committed:
[libcxx][docs] Make test name pattern documentation more obvious

As a new contributor, I found it hard to find the documentation for the
meaning of the names of different tests and how those names translate
to Lit.

This makes that documentation more explicit.

ldionne commented on libcxx/docs/TestingLibcxx.rst:
```diff
@@ -330,13 +330,48 @@ additional headers.
-``libcxx/utils/libcxx/test/format.py`` has documentation about writing test. It
-explains the difference between the test named  ``foo.pass.cpp`` and named
-``foo.verify.cpp`` are.
+Test names
```
I would like to avoid duplicating this documentation. It will get out of date when we add new test formats in the future.

I would suggest either migrating the whole documentation for our test format into `.rst` or dropping this change.

hawkinsw replied:
Okay!! Honestly, you tell me!! Which one would you prefer? I am happy to migrate it all and do a cleanup if you think it would be helpful!

ldionne replied:
I think documenting the libc++ testing format in `.rst` would be more helpful to contributors than it staying hidden where it's currently at. I've had to show it to people several times in the past so I think this is a common one that people don't know about. I would still link to these docs from `format.py` though just to connect them in an obvious way.

hawkinsw commented:
Great! Thank you for the "go ahead". I will send along some fixups and re-request review when it is ready!

hawkinsw committed:
fixup! [libcxx][docs] Make test name pattern documentation more obvious

Update to consolidate lit test file name pattern semantic documentation to docs/

hawkinsw commented:
I have updated to consolidate all documentation about the file name pattern semantics to `docs/TestingLibcxx.rst`. In the future, I would like to expand the testing documentation to include information about how to write a *builtin Lit shell* test (including the valid commands, how to use pipelines, valid substitutions, etc). @ldionne : If you would prefer to wait for that additional documentation before merging this PR, that is great by me. If not, I will make a separate PR.

mordante commented on libcxx/docs/TestingLibcxx.rst:
```diff
@@ -330,13 +330,48 @@ additional headers.
+
+The names of test files have meaning for the libcxx-specific configuration of 
+Lit. A complete description can be found in the docstring for the function 
+``CxxStandardLibraryTest`` in the file ``libcxx/utils/libcxx/test/format.py``.
```
The last sentence seems no longer to be true.

hawkinsw replied:
Good point! :-)

hawkinsw replied:
Sorry! I had stupidly not committed a change that fixed this!! Should be resolved now!

mordante commented on libcxx/utils/libcxx/test/format.py:
```diff
@@ -155,38 +155,10 @@ class CxxStandardLibraryTest(lit.formats.FileBasedTest):
-
+    Lit tests are contained in files that follow a certain pattern. That
+    pattern determines the semantics of the test. See
+    libcxx/docs/TestingLibcxx.rst or https://libcxx.llvm.org/TestingLibcxx.html
```
I would only link to the website and use an anchor.

hawkinsw replied:
Good point! Will do!

hawkinsw replied:
All done! Great feedback, thank you!

mordante reviewed (approved):
Thanks for working on this, I have grepped for this information a number times. Having it on the website gives it a better visibility.

LGTM modulo some comments. Please wait for @ldionne to review this too.

hawkinsw commented:
> Thanks for working on this, I have grepped for this information a number times. Having it on the website gives it a better visibility.
> 
> LGTM modulo some comments. Please wait for @ldionne to review this too.

Thank you for the feedback! I really appreciate it!

hawkinsw committed:
fixup! [libcxx][docs] Make test name pattern documentation more obvious

(Fully) centralize all naming documentation in the rst-based documentation.

hawkinsw committed:
fixup! [libcxx][docs] Make test name pattern documentation more obvious

Point only to www documentation for test names (per @mordante's excellent feedback)

ldionne commented on libcxx/utils/libcxx/test/format.py:
```diff
@@ -155,38 +155,10 @@ class CxxStandardLibraryTest(lit.formats.FileBasedTest):
-                              automatically marked as UNSUPPORTED if the compiler
-                              does not support Clang-verify.
-
+    Lit tests are contained in files that follow a certain pattern. That
+    pattern determines the semantics of the test. See
+    https://libcxx.llvm.org/TestingLibcxx.html#test-names
+    for a complete description of those semantics.
```
```suggestion
    Lit tests are contained in files that follow a certain pattern, which determines the semantics of the test.
    Under the hood, we basically generate a builtin Lit shell test that follows the ShTest format, and perform
    the appropriate operations (compile/link/run). See
    https://libcxx.llvm.org/TestingLibcxx.html#test-names
    for a complete description of those semantics.
```

ldionne commented on libcxx/utils/libcxx/test/format.py:
```diff
@@ -155,38 +155,10 @@ class CxxStandardLibraryTest(lit.formats.FileBasedTest):
+    https://libcxx.llvm.org/TestingLibcxx.html#test-names
+    for a complete description of those semantics.
 
     Substitution requirements
```
As a follow-up, it would be great to also move this part of the documentation to `.rst` so that all of the documentation for the file format is in the same place.

ldionne reviewed (approved):
LGTM w/ comments. It would be nice to move the rest of the documentation for the test format to `.rst` too, otherwise we'll have things in multiple places which means more likely to forget to update new information everywhere.

hawkinsw commented:
> LGTM w/ comments. It would be nice to move the rest of the documentation for the test format to `.rst` too, otherwise we'll have things in multiple places which means more likely to forget to update new information everywhere.

Thank you for the review, @ldionne !! I will incorporate your feedback ASAP! The other thing that I would like to add with this PR is documentation on the actual commands available in the lit built-in shell tests. Does that seem reasonable? If so, I will start working on that and try to have something ready for tomorrow!

Thank you again!

ldionne commented:
> The other thing that I would like to add with this PR is documentation on the actual commands available in the lit built-in shell tests.

Which commands do you mean? The actual builtin Lit ShTest commands like `RUN`? If so, those are already documented in Lit-specific documentation. We should move the documentation for the commands we add in our test format (like `ADDITIONAL_COMPILE_FLAGS`), though.

hawkinsw commented:
> > The other thing that I would like to add with this PR is documentation on the actual commands available in the lit built-in shell tests.
> 
> Which commands do you mean? The actual builtin Lit ShTest commands like `RUN`? If so, those are already documented in Lit-specific documentation. We should move the documentation for the commands we add in our test format (like `ADDITIONAL_COMPILE_FLAGS`), though.

Great! I will work on that. 

I was talking about documenting the builtin lit "shell" commands (e.g., 

```Python
    inproc_builtins = {
        "cd": executeBuiltinCd,
        "export": executeBuiltinExport,
        "echo": executeBuiltinEcho,
        "@echo": executeBuiltinEcho,
        "mkdir": executeBuiltinMkdir,
        "popd": executeBuiltinPopd,
        "pushd": executeBuiltinPushd,
        "rm": executeBuiltinRm,
        ":": executeBuiltinColon,
    }   
```

and if/how they differ in their meaning from posix (or PowerShell). Those are not documented in the other lit documentation (e.g., https://llvm.org/docs/TestingGuide.html#writing-new-regression-tests) and it was something that I struggled to find. Would you prefer that I send that as PR directly to lit?

I will reflag you for a review when it is ready! Thanks again!

hawkinsw committed:
Update libcxx/docs/TestingLibcxx.rst

Be more clear about what type of shell we get when running these tests.

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

hawkinsw committed:
Update libcxx/utils/libcxx/test/format.py

Make it clearer how the tests are generated by libcxx to drive lit according to the patterns embedded in the filenames that contain tests.

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

hawkinsw committed:
fixup! [libcxx][docs] Make test name pattern documentation more obvious

Add documentation about ADDITIONAL_COMPILE_FLAGS and FILE_DEPENDENCIES.

hawkinsw commented:
> > The other thing that I would like to add with this PR is documentation on the actual commands available in the lit built-in shell tests.
> 
> Which commands do you mean? The actual builtin Lit ShTest commands like `RUN`? If so, those are already documented in Lit-specific documentation. We should move the documentation for the commands we add in our test format (like `ADDITIONAL_COMPILE_FLAGS`), though.

I have added documentation for the `ADDITIONAL_COMPILE_FLAGS` and `FILE_DEPENDENCIES` directives.

hawkinsw commented on libcxx/docs/TestingLibcxx.rst:
```diff
@@ -330,13 +330,53 @@ additional headers.
+   * - ``FOO.sh.<anything>``
+     - A *builtin Lit Shell* test.
+   * - ``FOO.gen.<anything>``
+     - A *builtin Lit Shell* test that generates one or more Lit tests on the fly. Executing this test must generate one or more files as expected
```
Thank you! I have committed these!

ldionne commented on libcxx/docs/TestingLibcxx.rst:
```diff
@@ -330,13 +330,83 @@ additional headers.
+       specified using relative paths will be anchored to the directory specified by the ``%S`` substitution (i.e., the source directory of the test being executed.).
+   * - ``ADDITIONAL_COMPILE_FLAGS``
+     - ``// ADDITIONAL_COMPILE_FLAGS: flag1, flag2, ...``
+     - The additional compiler flags specified using the ``ADDITIONAL_COMPILE_FLAGS`` libc++-specific Lit directive will be added to the invocation of ``clang`` used to build
```
This new documentation is not accurate. They don't apply to Clang only, and they do affect `.sh.cpp` tests too. I suggest we stick to the previous documentation:

```
This directive will cause the provided flags to be added to the
%{compile_flags} substitution for the test that contains it. This
allows adding special compilation flags without having to use a
.sh.cpp test, which would be more powerful but perhaps overkill.
```

hawkinsw replied:
I believe that I have properly incorporated this feedback as well! Thanks!

ldionne commented on libcxx/docs/TestingLibcxx.rst:
```diff
@@ -330,13 +330,83 @@ additional headers.
+     - Usage
+   * - ``FILE_DEPENDENCIES``
+     - ``// FILE_DEPENDENCIES: file, directory, /path/to/file, ...``
+     - The paths given to the ``FILE_DEPENDENCIES`` directive can specify directories or specific files upon which a given test depend. Copies of the files individually specified and
```
This is slightly incorrect. This uses the `%T` substitution, not `%S`. I suggest keeping the wording in the old documentation, I think it wasn't unclear.

hawkinsw replied:
The reason I wrote `%S` was 

```
# For each file dependency in FILE_DEPENDENCIES, inject a command to copy
# that file to the execution directory. Execute the copy from %S to allow
# relative paths from the test directory.
```

from line 130 of `format.py`. Sorry!

ldionne replied:
Ah, I see -- you're right, this was missing from the old documentation. Maybe something like this?

```
This directive expresses that the test requires the provided files
or directories in order to run. An example is a test that requires
some test input stored in a data file. When a test file contains
such a directive, the files will collect them and copy them
to the directory represented by %T. The copy is performed from
the directory represented by %S (i.e. the source directory of  the
test being executed), which allows using relative paths to specify
files. After performing the copy, the intent is that %T contains all
the inputs necessary to run the test, such that e.g. execution on a
remote host can be done by simply copying %T to the host.
```

hawkinsw replied:
Let me know what you think about what I am about to commit. I think that your suggestion was great -- the examples really help!

hawkinsw commented:
@ldionne Thank you for the feedback. I am sorry that there were mistakes. I am *sure* that the mistakes were all my fault but I didn't intend to change any of the documentation that was previously written. I will address your comments as soon as possible. I am going to be away from my keyboard until Thursday but will work on it ASAP.

hawkinsw commented on libcxx/utils/libcxx/test/format.py:
```diff
@@ -155,38 +155,11 @@ class CxxStandardLibraryTest(lit.formats.FileBasedTest):
-    FOO.verify.cpp          - Compiles with clang-verify. This type of test is
-                              automatically marked as UNSUPPORTED if the compiler
-                              does not support Clang-verify.
-
```
All done!!

hawkinsw commented on libcxx/docs/TestingLibcxx.rst:
```diff
@@ -330,13 +330,83 @@ additional headers.
+       *all* the files in any specified directories will be placed in the *current working directory* of the test when it executes. All dependency directories and files
+       specified using relative paths will be anchored to the directory specified by the ``%S`` substitution (i.e., the source directory of the test being executed.).
+   * - ``ADDITIONAL_COMPILE_FLAGS``
+     - ``// ADDITIONAL_COMPILE_FLAGS: flag1, flag2, ...``
```
I believe I got that in the latest fixup.

hawkinsw committed:
fixup! [libcxx][docs] Make test name pattern documentation more obvious

Address @ldionne's astute feedback.

hawkinsw commented:
@mordante In re: #76246 I have added some documentation for the `MODULE` directive. Let me know if it reads okay!

hawkinsw committed:
fixup! [libcxx][docs] Make test name pattern documentation more obvious

Add link to Internals Manual for writing clang-verify tests.

hawkinsw committed:
fixup! [libcxx][docs] Make test name pattern documentation more obvious

Add recommendation on when to write verify tests.

ldionne commented on libcxx/docs/TestingLibcxx.rst:
```diff
@@ -325,18 +325,98 @@ This macro is in a different header as ``assert_macros.h`` since it pulls in
+     - Same as ``FOO.pass.cpp``, but for Objective-C++.
+
+   * - ``FOO.compile.pass.cpp``
+     - Checks whether the C++ code in the file compiles successfully. In general, prefer `compile` tests over `verify` tests, 
```
I think you need to use double-backticks here for this to render properly!

cjdb replied:
I'm surprised this isn't a build error.

ldionne replied:
I think it's valid RST code, it will just render weird.

hawkinsw replied:
Of course I do!! Sorry. Thank you!

ldionne reviewed (approved):
LGTM with minor changes to satisfy RST files. Thanks!

hawkinsw committed:
fixup! [libcxx][docs] Make test name pattern documentation more obvious

Address final comments.

hawkinsw commented:
Can I get one more check, @ldionne ? Thanks -- I'm just a little nervous!

ldionne reviewed (approved):
LGTM
